src/core/net/srp_server.cpp

```diff
@@ -214,11 +214,41 @@ void Server::AddHost(Host *aHost)
     IgnoreError(mHosts.Add(*aHost));
 }
 
-void Server::RemoveAndFreeHost(Host *aHost)
+void Server::RemoveHost(Host *aHost, bool aRetainName, bool aNotifyServiceHandler)
 {
-    otLogInfoSrp("[server] fully remove host %s", aHost->GetFullName());
-    IgnoreError(mHosts.Remove(*aHost));
-    aHost->Free();
+    VerifyOrExit(aHost != nullptr);
+
+    aHost->mLease = 0;
+    aHost->ClearResources();
+
+    if (aRetainName)
+    {
+        otLogInfoSrp("[server] remove host '%s' (but retain its name)", aHost->mFullName);
+    }
+    else
+    {
+        aHost->mKeyLease = 0;
+        IgnoreError(mHosts.Remove(*aHost));
+        otLogInfoSrp("[server] fully remove host '%s'", aHost->mFullName);
+    }
+
+    if (aNotifyServiceHandler && mServiceUpdateHandler != nullptr)
+    {
+        mServiceUpdateHandler(aHost, kDefaultEventsHandlerTimeout, mServiceUpdateHandlerContext);
```

Comment:
Is it safe for the SRP Server observer to call `otSrpServerHandleServiceUpdateResult()` in the handler? Is possible to run into a "double free" issue on `aHost` below?

Reply:
I think @wgtdkp knows this better and can confim this
But I guess/think it should be safe since in `UpdateResult ` we seem to check in `mOutstandingUpdates` list and here we are not adding anything to `mOutstandingUpdates` list. So the `UpdatResult` (if called) should do nothing.

Reply:
Thanks @abtink, yes it is safe for the observer to call `otSrpServerHandleServiceUpdateResult` and it will be ignored.
We have the invariant that a host is either in the `OutstandingUpdates` or in the `mHosts` list.

But I am feeling that using the `Host` pointer to identify a SRP update transaction is not safe. For, example, a new Host object may be allocated with the same address of an already freed host object and a callback for the freed host will match the newly allocated host object. 

I have been planning to use an `uint32_t` SRP update/transaction ID that can uniquely identify a SRP update but did't get a chance yet. I will have the change ASAP.

Reply:
> I have been planning to use an uint32_t SRP update/transaction ID that can uniquely identify a SRP update. Thoughts?

This sounds like a very good idea to me. 👍